### PR TITLE
Add Char * String and Char * Char

### PR DESCRIPTION
**Pull request by adamslc**
Solves #22512. I am definitely a Github novice, so let me know if I made a stupid mistake somewhere...

### CODE REVIEW

**Comment by KristofferC** [comment]
Perhaps you could also add some tests of the kind `'a' * "b" * 'c'` etc?

**Comment by adamslc** [comment]
I've added a few more tests. I tried to squash the extra commit, but I just made a horrible mess. How should I do that?

**Comment by KristofferC** [comment]
It's OK, it can be squashed on merge with the github ui.

File: base/strings/basic.jl [hunk 1]
```diff
@@ -68,6 +68,9 @@ julia> "Hello " * "world"
 ```
 """
 (*)(s1::AbstractString, ss::AbstractString...) = string(s1, ss...)
+(*)(c::Char, s::AbstractString) = string(c, s)
+(*)(s::AbstractString, c::Char) = string(s, c)
+(*)(c1::Char, c2::Char) = string(c1, c2)
```
Comment (pabloferz) [inline base/strings/basic.jl]: We might subsume all these with 

```julia
(*)(s::AbstractString, ss::Union{Char,AbstractString}...) = string(s, ss...)
(*)(c::Char, ss::Union{Char,AbstractString}...) = string(c, ss...)
```

and take advantage of the `string(::Union{Char,AbstractString}...)` method.

**Comment by andyferris** [comment]
I was wondering about concatenating characters to make this operation more complete, e.g. `char * char` and things like `'a' * 'b' * "c"` vs `"a" * 'b' * 'c'`.

**Comment by andyferris** [comment]
PS `++` rules 😜

File: base/strings/basic.jl (hunk 1 above)
Comment (adamslc) [inline base/strings/basic.jl]: I reduced it further to one method (see next commit)

**Comment by tkelman** [comment]
docstring should be updated here - it isn't formatted correctly on master, signatures should be indented, not backtick fenced

File: base/strings/basic.jl [hunk 2]
```diff
@@ -56,15 +56,16 @@ sizeof(s::AbstractString) = error("type $(typeof(s)) has no canonical binary rep
 eltype(::Type{<:AbstractString}) = Char
 
 """
-```
-*(s::AbstractString, t::AbstractString)
-```
+    *(s::Union{Char, AbstractString}, t::Union{Char, AbstractString})
```
Comment (tkelman) [inline base/strings/basic.jl]: ... on the second input

**Comment by adamslc** [comment]
Does this need anything else?

**Comment by pabloferz** [comment]
There was a timeout on linux 32-bit so I restated the build (gist backed-up here https://gist.github.com/pabloferz/4cb263e0967ba9a5c3256bc53d3619ee)

**Review by KristofferC — APPROVED** [review]

File: base/strings/basic.jl [hunk 3]
```diff
@@ -56,18 +56,19 @@ sizeof(s::AbstractString) = error("type $(typeof(s)) has no canonical binary rep
 eltype(::Type{<:AbstractString}) = Char
 
 """
-```
-*(s::AbstractString, t::AbstractString)
-```
+    *(s::Union{Char, AbstractString}, t::Union{Char, AbstractString}...)
 
-Concatenate strings. The `*` operator is an alias to this function.
+Concatenate strings and characters. The `*` operator is an alias to this function.
```
Comment (fredrikekre) [inline base/strings/basic.jl]: "[...] and characters to a ``[`String`](@ref)``." perhaps?

Also, this is the `*` function, its an alias for itself!?

Comment (ararslan) [inline base/strings/basic.jl]: I would make it like this:
```julia
"""
    *(s::Union{AbstractString, Char}, t::Union{AbstractString, Char})

Concatenate strings and/or characters, producing a [`String`](@ref). This is equivalent
to calling the [`string`](@ref) function on the arguments.
"""
```

**Review by ararslan — APPROVED** [review]

File: base/strings/basic.jl [hunk 4]
```diff
@@ -56,9 +56,12 @@ sizeof(s::AbstractString) = error("type $(typeof(s)) has no canonical binary rep
...
 
 """
-    *(s::Union{Char, AbstractString}, t::Union{Char, AbstractString}...)
+    *(s::Union{AbstractString, Char}, t::Union{AbstractString, Char})
 
-Concatenate strings and characters. The `*` operator is an alias to this function.
+Concatenate strings and/or characters, producing a [`String`](@ref). This is equivalent
+to calling the [`string`](@ref) function on the arguments.
+
+# Examples
```
Comment (tkelman) [inline base/strings/basic.jl]: delete the blank line

Comment (ararslan) [inline base/strings/basic.jl]: I thought we had been putting lines between the headers and contents in docstring?

Comment (ararslan) [inline base/strings/basic.jl]: Ah, I guess not.

Comment (tkelman) [inline base/strings/basic.jl]: recently moving towards getting rid of them everywhere

Comment (ararslan) [inline base/strings/basic.jl]: K, thanks for the heads up

File: base/strings/basic.jl [hunk 5]
```diff
@@ -59,20 +59,21 @@ sizeof(s::AbstractString) = error("type $(typeof(s)) has no canonical binary rep
...
 ```jldoctest
 julia> "Hello " * "world"
 "Hello world"
+
+julia> 'j' * "ulia"
+"julia"
 ```
 """
-(*)(s1::AbstractString, ss::AbstractString...) = string(s1, ss...)
+(*)(s1::Union{Char, AbstractString}, ss::Union{Char, AbstractString}...) = string(s1, ss...)
```
Comment (musm) [inline base/strings/basic.jl]: Aside from some of the linalg code, typically there isn't a space after the comma in `Union`. Probably best to keep consistency within this file and alike.

**Comment by musm** [comment]
bump

**Comment by StefanKarpinski** [comment]
Unless there are objections, I plan to merge this in 24 hours.

File: base/strings/basic.jl [hunk 6]
```diff
@@ -56,9 +56,12 @@ sizeof(s::AbstractString) = error("type $(typeof(s)) has no canonical binary rep
 eltype(::Type{<:AbstractString}) = Char
 
 """
-    *(s::Union{Char, AbstractString}, t::Union{Char, AbstractString}...)
+    *(s::Union{AbstractString, Char}, t::Union{AbstractString, Char})
```
Comment (pabloferz) [inline base/strings/basic.jl]: Why where the three dots removed?

**Comment by ararslan** [comment]
Once the `...` is added back to the docstring, LGTM.

**Comment by KristofferC** [comment]
Just add it back (with collaborator access to the branch) and merge?

**Comment by ararslan** [comment]
Good idea, @KristofferC. Done.

**Comment by ararslan** [comment]
Thanks for the contribution, @adamslc! Nice work here.

**Comment by stevengj** [comment]
* Since it defines `Char * Char = string`, it should also define `one(::Type{Char}) = ""`.

* Should probably have a specialized `Char^Integer` method similar to the `String^Integer` method.

**Comment by ararslan** [comment]
Regarding `one`, isn't it typically assumed that `one(::Type{T})` has type `T`? Defining that method for `Char` would break that, and AFAIK it would be the only exception. (There may be others that I don't know about though.)

**Comment by stevengj** [comment]
@ararslan, no, that is not correct.  e.g. `one` for a dimensionful quantity returns a different type.  (If you want the same type, you call `oneunit`, which isn't defined here.)

**Comment by stevengj** [comment]
We should definitely have a specialized `^`, however.  The default one isn't type-stable for `Char` and is grossly inefficient.

**Comment by musm** [comment]
@stevengj 

The following specialized version of `repeat`, which `^` calls seems to work fine.
```julia

function repeat(s::Char, r::Integer)
    r < 0 && throw(ArgumentError("can't repeat a string $r times"))
    out = _string_n(r)
    ccall(:memset, Ptr{Void}, (Ptr{UInt8}, Cint, Csize_t), out, s, r)
    return out
end

```
There isn't much speed improvement over `repeat(s::String,r::Integer)

**Comment by stevengj** [comment]
(That only works for ascii chars.  For non-ascii I would just call `string(s)^r` as a fallback.)

**Comment by musm** [comment]
Calling `repeat(string(s), 3)` would allocate and makes it about twice as slow.

**Comment by stevengj** [comment]
@musm, I understand that, but since `char^integer` and `string^integer` are almost exclusively used for ASCII chars (mainly for repeating spaces), I think it is fine to optimize mainly the ASCII case and leave the non-ASCII case to a slower fallback for now.

**Comment by StefanKarpinski** [comment]
To implement an efficient character repeating operator, it's sufficient to figure out what 1-4 byte pattern the character produces in UTF-8 and then copy that as many times as the character needs to be repeated. Not entirely straightforward, but not crazy to implement either.

**Comment by ararslan** [comment]
Is there something relevant already implemented in utf8proc?

**Comment by stevengj** [comment]
Stefan, I know its possible, but I don't think it is worth the trouble

**Comment by StefanKarpinski** [comment]
Sure, can always be done as an optimization in the future some time.

**Comment by musm** [comment]
@ScottPJones  sent me the following version which does not allocate a while back. I don't think he has had the time to open a PR on his branch so I am posting this here in the hopes that someone opens a PR with the change

```julia
function repeat(c::Char, r::Integer)
    r < 0 && throw(ArgumentError("can't repeat a character $r times"))
    r == 0 && return ""
    ch = UInt(c)
    if ch < 0x80
        out = Base._string_n(r)
        ccall(:memset, Ptr{Void}, (Ptr{UInt8}, Cint, Csize_t), out, c, r)
    elseif ch < 0x800
        out = _string_n(2r)
        p16 = reinterpret(Ptr{UInt16}, pointer(out))
        u16 = ((ch >> 0x6) | (ch & 0x3f) << 0x8) % UInt16 | 0x80c0
        @inbounds for i = 1:r
            unsafe_store!(p16, u16, i)
        end
    elseif ch < 0x10000
        (0xd800 ≥ ch ≤ 0xdfff) || throw(ArgumentError("invalid character 0x$(hex(ch))"))
        out = _string_n(3r)
        p = pointer(out)
        b1 = (ch >> 0xc) % UInt8 | 0xe0
        b2 = ((ch >> 0x6) & 0x3f) % UInt8 | 0x80
        b3 = (ch & 0x3f) % UInt8 | 0x80
        @inbounds for i = 1:r
            unsafe_store!(p, b1)
            unsafe_store!(p, b2, 2)
            unsafe_store!(p, b3, 3)
            p += 3
        end
    elseif ch < 0x110000
        out = _string_n(4r)
        p32 = reinterpret(Ptr{UInt32}, pointer(out))
        u32 = ((ch >> 0x12) | ((ch >> 0x4) & 0x03f00) |
            ((ch << 0xa) & 0x3f0000) | ((ch & 0x3f) << 0x18)) % UInt32 | 0x808080f0
        @inbounds for i = 1:r
            unsafe_store!(p32, u32)
            p32 += 4
        end
    else
        throw(ArgumentError("invalid character 0x$(hex(ch))"))
    end
    return out
end
```

**Comment by StefanKarpinski** [comment]
Thanks, PR: https://github.com/JuliaLang/julia/pull/23787.

**Pull request merged**